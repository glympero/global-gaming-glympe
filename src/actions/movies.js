//SET_MOVIES
const setMovie = (movies) => ({
    type: 'SET_MOVIES',
    movies
  });

// REMOVE_MOVIE
const removeMovie = ({ id } = {}) => ({
    type: 'REMOVE_MOVIE',
    id
});

// EDIT_MOVIE
const editMovie = (id, updates) => ({
    type: 'EDIT_MOVIE',
    id,
    updates
});

export { setMovie, removeMovie, editMovie };