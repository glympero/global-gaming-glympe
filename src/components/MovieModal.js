import React from 'react';
import PropTypes from 'prop-types';
import { Dialog } from 'material-ui';
import Edit from "material-ui/svg-icons/image/edit"
import moment from 'moment';

const styles = {
    dialogContent: (backgroundUrl) => ({
      backgroundImage: `linear-gradient(rgba(0, 0, 0, 0.6), rgba(0, 0, 0, 0.6)), url(${backgroundUrl})`,
      backgroundRepeat: 'no-repeat',
      backgroundSize: '100%',
      height: '100%',
      minHeight: 400,
      color: 'white',
      padding: 10
    })
  }

class MovieForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      title: props.movie ? props.movie.title : '',
      allowEdit: false,
      error: ''
    }
  }

  onTitleChange = (e) => {
    const title = e.target.value;
    this.setState(() => ({
      title
    }));
  }
  
  onSubmit = (e) => {
    e.preventDefault();
    if(
      !this.state.title
    ) {
      this.setState(() => ({
        error: 'Title is required'
      }));
    }else {
      this.setState(() => ({
        error: ''
      }));
      const title = this.state.title;
      this.props.onSubmit({ 
        title,
      });
    }
  }

  onAllowEditChange = () => {
    this.setState({
      allowEdit: !this.state.allowEdit
    })
  }

  confirmRemove = () => {
    const { remove } = this.props;
    const result = confirm("Want to delete?");
    if(result){
      remove();
    }
    return false;
  }

render() {
    const { movie } = this.props;
    const { allowEdit, title, error } = this.state;
    return (
      <Dialog
        autoScrollBodyContent={true}
        title={null}
        modal={false}
        open={true}
      >
        <div style={styles.dialogContent(movie.poster_path)}>
        <h1>{title} <button className='editButton' onClick={this.onAllowEditChange}><Edit style={{ color: 'red'}}/></button></h1>
        { allowEdit && <input className='text-input' type='text' value={title} onChange={this.onTitleChange} placeholder='title' autoFocus/> }
        { !!error && <p className='form__error'>{error}</p>}
        <h5>{movie.genre}</h5>
        <p>{movie.description}</p>
        <p>Director: {movie.director.name}</p>
        <p>Release Date: {moment(movie.release_date, "Do MMM YYYY")}</p>
        <div className='modal--buttons'>
          <button className='button button--modal' onClick={this.onSubmit}>Close</button>
          { allowEdit && <button className='button button--modal button--danger' onClick={this.confirmRemove}>Remove</button> }
        </div>
        </div>
    </Dialog>
    );
  }
}

MovieForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  movie: PropTypes.object.isRequired,
  remove: PropTypes.func.isRequired
};

export default MovieForm;