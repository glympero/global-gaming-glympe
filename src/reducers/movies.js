const moviesReducerDefaultState = [];

const moviesReducer = (state = moviesReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_MOVIES':
            return action.movies;
        case 'REMOVE_MOVIE':
            return state.filter(movie => movie.id !== action.id);
        case 'EDIT_MOVIE':
            return state.map(movie => {
                if(movie.id === action.id) {
                    return {
                        ...movie,
                        ...action.updates
                    }
                }
                return movie;
            })
        default:
            return state;
    }
};

export default moviesReducer;