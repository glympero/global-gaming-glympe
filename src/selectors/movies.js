const getVisibleMovies = (movies, filters) => {
    const {text, sortBy, genre } = filters;
    return movies.filter(movie => {
        const textMatched = typeof text === 'string' && movie.title.toLowerCase().indexOf(text.toLowerCase()) !== -1;
        const genreMatched = typeof genre === 'string' && genre === 'all' ? true : movie.genre.toLowerCase().indexOf(genre.toLowerCase()) !== -1;
        return textMatched && genreMatched;
    }).sort((a, b) => {
        if(sortBy === 'title') {
            return a.title > b.title ? 1 : -1;
        }else if (sortBy === 'director') {
            return a.director.toLowerCase() > b.director.toLowerCase() ? 1 : -1;
        }
    });
};

export default getVisibleMovies;