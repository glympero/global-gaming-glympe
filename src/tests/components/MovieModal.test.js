import React from 'react';
import { shallow } from 'enzyme';
import MovieModal from '../../components/MovieModal';
import movies from '../../fixtures/data';

let onSubmit, wrapper, match, remove
beforeEach(() => {
    onSubmit = jest.fn();
    match = { params: 1 }
    remove = jest.fn();

    wrapper = shallow(
        <MovieModal 
            onSubmit={onSubmit} 
            match={match}
            movie={movies[0]}
            remove={remove}
        />
    )
});

test('should render MovieModal correctly', () => {
    expect(wrapper).toMatchSnapshot();
});