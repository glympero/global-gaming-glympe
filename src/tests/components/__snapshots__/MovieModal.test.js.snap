// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render MovieModal correctly 1`] = `
<Dialog
  autoDetectWindowHeight={true}
  autoScrollBodyContent={true}
  modal={false}
  open={true}
  repositionOnUpdate={true}
  title={null}
>
  <div
    style={
      Object {
        "backgroundImage": "linear-gradient(rgba(0, 0, 0, 0.6), rgba(0, 0, 0, 0.6)), url(12345)",
        "backgroundRepeat": "no-repeat",
        "backgroundSize": "100%",
        "color": "white",
        "height": "100%",
        "minHeight": 400,
        "padding": 10,
      }
    }
  >
    <h1>
      Absolute Bite
       
      <button
        className="editButton"
        onClick={[Function]}
      >
        <ImageEdit
          style={
            Object {
              "color": "red",
            }
          }
        />
      </button>
    </h1>
    <h5>
      Drama@april.biz
    </h5>
    <p />
    <p>
      Director: 
      Walsh Mertie
    </p>
    <p>
      Release Date: 
    </p>
    <div
      className="modal--buttons"
    >
      <button
        className="button button--modal"
        onClick={[Function]}
      >
        Close
      </button>
    </div>
  </div>
</Dialog>
`;
